pkgs/development/python-modules/gpuctypes/default.nix

```diff
@@ -0,0 +1,63 @@
+{ lib
+, config
+, buildPythonPackage
+, fetchFromGitHub
+, substituteAll
+, addDriverRunpath
+, cudaSupport ? config.cudaSupport
+, rocmSupport ? config.rocmSupport
+, cudaPackages
+, setuptools
+, ocl-icd
+, rocmPackages
+}:
+buildPythonPackage rec {
+  pname = "gpuctypes";
+  version = "0.3.0";
+  pyproject = true;
+
+  src = fetchFromGitHub {
+    repo = "gpuctypes";
+    owner = "tinygrad";
+    rev = "refs/tags/${version}";
+    hash = "sha256-xUMvMBK1UhZaMZfik0Ia6+siyZGpCkBV+LTnQvzt/rw=";
+  };
+
+  patches = [
+    (substituteAll {
+      src = ./fix-dlopen-cuda.patch;
+      inherit (addDriverRunpath) driverLink;
+      libnvrtc =
+        if cudaSupport
+        then "${lib.getLib cudaPackages.cuda_nvrtc}/lib/libnvrtc.so"
+        else "Please import nixpkgs with `config.cudaSupport = true`";
+    })
+  ];
+
+  nativeBuildInputs = [
+    setuptools
+  ];
+
+  postPatch = ''
+    substituteInPlace gpuctypes/opencl.py \
+      --replace "ctypes.util.find_library('OpenCL')" "'${ocl-icd}/lib/libOpenCL.so'"
+  ''
+  + lib.optionalString rocmSupport ''
+    substituteInPlace gpuctypes/hip.py \
+      --replace "/opt/rocm/lib/libamdhip64.so" "${rocmPackages.clr}/lib/libamdhip64.so"
+    substituteInPlace gpuctypes/hip.py \
+      --replace "/opt/rocm/lib/libhiprtc.so" "${rocmPackages.clr}/lib/libhiprtc.so"
+  '';
+
+  pythonImportsCheck = [ "gpuctypes" ];
+
+  # All tests require a GPU to work. This is not possible in the sandbox.
```

Comment:
This isn't quite true, since a lot of the tests can be run with a different backend, like "CLANG", "LLVM", or "CPU"

Reply:
https://github.com/tinygrad/gpuctypes/tree/master/test

![image](https://github.com/NixOS/nixpkgs/assets/9720532/5049eb57-5355-45b8-8675-0d33609012c6)

Hm?

Reply:
Oh my bad read package name wrong, thought this was in tinygrad

Reply:
But the compilation tests should still be possible to run since they should need a GPU to be present: https://github.com/tinygrad/gpuctypes/blob/c4c4394239dce4d4ecfe7016ca268c49cb2d02a4/test/test_hip.py#L19

Reply:
We tried to run the OpenCL ones but got runtime error... Do you think that `test_hip.py` has better chances ?

Reply:
The opencl ones do require a device hip and cuda, not the device tests should be runnable

Reply:
After blackmagic-patching, @SomeoneSerge and I were able to run the compilation tests for HIP and CUDA !